url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/training/GradientDescent.java

```diff
@@ -0,0 +1,489 @@
+    
+    /**
+     * Resets and current weight differences and stores them
+     * in history memory.
+     */
+    private void resetWeightDiffs() {
+        //
+        // initialize weight diffs.
+        //
+        for (int i = 1; i <= this.weightsnum; i++) {
+            this.dweightslast[i] = this.dweights[i];
+            this.dweights[i]     = 0.0;
+        }
+    }
+    
+    /**
+     * Accumulates the weight differences for the given frame index (time index) 
+     * and the current gradient results.
+     * @param frameidx Frame index (time index).
+     */
+    private void accumulateWeightsDiffs(final int frameidx) {
+        //
+        // collect deltas * activations .
+        //
+        for (int t = 0; t <= frameidx; t++) {
+            //
+            final double[] outputs = this.net.getOutputBuffer(t);
+            final double[] deltas  = this.net.getGradOutputBuffer(t);
+            //
+            /*
+            // # DEBUG #
+            final double[] deltasin = net.getGradInputBuffer(t);
+            for (int i = 0; i < deltas.length; i++) {
+                this.deltasdbg[i] += Math.abs(deltas[i]);
+                this.deltasindbg[i] += Math.abs(deltasin[i]);
+            }
+            // #
+            */
+            int off = 0;
+            //
+            for (int i = 0; i < this.linksnum; i++) {
+                //
+                final int src  = this.links[off + Link.IDX_SRC];
+                final int dst  = this.links[off + Link.IDX_DST];
+                final int widx = this.links[off + Link.IDX_WEIGHT];
+                //
+                if (widx > 0) {
+                    final double dk = deltas[dst];
+                    final double xj = outputs[src];
+                    //                     
+                    final double dw = (dk * xj);
+                    this.dweights[widx] += (dw);
+                }
+                //
+                off += Link.LINK_SIZE;
+            }
+        }
+    }
+    
+    /**
+     * Finally computes the weight differences based on the accumulated
+     * values.
+     */
+    private void computeWeightDiffs() {
+        //
+        // add momentum and multiply learning rate mu. 
+        //
+        for (int i = 1; i <= this.weightsnum; i++) {
+            //
+            // calculate momentum.
+            //
+            final double mw = this.dweightslast[i] * this.alpha; 
+            this.dweights[i] = (this.mu * this.dweights[i]) + mw;
+        }  
+    }
+    
+    /**
+     * Adjusts the weights.
+     */
+    private void adjustWeights() {
+        //
+        for (int i = 1; i <= this.weightsnum; i++) {
+            //
+            this.weights[i] += this.dweights[i];
+            /*
+            // # DEBUG #
+            this.weightsdbg[i] += Math.abs(this.dweights[i]);
+            // #
+            */
+        }
+        
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    synchronized
+    public void train() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.